examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,2035 @@
+            dirs = sorted(dirs, key=lambda x: int(x.split("-")[1]))
+            path = dirs[-1] if len(dirs) > 0 else None
+
+        if path is None:
+            accelerator.print(
+                f"Checkpoint '{args.resume_from_checkpoint}' does not exist. Starting a new training run."
+            )
+            args.resume_from_checkpoint = None
+            initial_global_step = 0
+        else:
+            accelerator.print(f"Resuming from checkpoint {path}")
+            accelerator.load_state(os.path.join(args.output_dir, path))
+            global_step = int(path.split("-")[1])
+
+            initial_global_step = global_step
+            first_epoch = global_step // num_update_steps_per_epoch
+    else:
+        initial_global_step = 0
+
+    progress_bar = tqdm(
+        range(0, args.max_train_steps),
+        initial=initial_global_step,
+        desc="Steps",
+        # Only show the progress bar once on each machine.
+        disable=not accelerator.is_local_main_process,
+    )
+
+    for epoch in range(first_epoch, args.num_train_epochs):
+        for step, batch in enumerate(train_dataloader):
+            with accelerator.accumulate(unet):
+                # 1. Load and process the image and text conditioning
+                if args.use_image_conditioning:
+                    image, text, cond_image = batch
+                else:
+                    image, text = batch
+
+                image = image.to(accelerator.device, non_blocking=True)
+                encoded_text = compute_embeddings_fn(text)
+                if args.use_image_conditioning:
+                    encoded_image = compute_image_embeddings_fn(cond_image)
+
+                pixel_values = image.to(dtype=weight_dtype)
+                if vae.dtype != weight_dtype:
+                    vae.to(dtype=weight_dtype)
+
+                # encode pixel values with batch size of at most args.vae_encode_batch_size
+                latents = []
+                for i in range(0, pixel_values.shape[0], args.vae_encode_batch_size):
+                    latents.append(vae.encode(pixel_values[i : i + args.vae_encode_batch_size]).latent_dist.sample())
+                latents = torch.cat(latents, dim=0)
+
+                latents = latents * vae.config.scaling_factor
+                latents = latents.to(weight_dtype)
+                bsz = latents.shape[0]
+
+                # 2. Sample random student timesteps s uniformly in `student_timestep_schedule` and sample random
+                # teacher timesteps t uniformly in [0, ..., noise_scheduler.config.num_train_timesteps - 1].
+                student_index = torch.randint(0, student_distillation_steps, (bsz,), device=latents.device).long()
+                student_timesteps = student_timestep_schedule[student_index]
+                teacher_timesteps = torch.randint(
+                    0, noise_scheduler.config.num_train_timesteps, (bsz,), device=latents.device
```

Comment:
it might better not to sample `teacher_timesteps` from the full range of `[0, noise_scheduler.config.num_train_timesteps)` but instead ignoring timesteps that are too small or too big, e.g. the default configurable range from dream fusion is `[0.02, 0.98] * num_train_timesteps`

Reply:
I guess Figure 2 of the [ADD paper](https://arxiv.org/pdf/2311.17042.pdf) implies that they sample from the full range of teacher timesteps:

<img width="310" alt="add_figure_2_teacher_timesteps" src="https://github.com/huggingface/diffusers/assets/58458699/c5a75e1e-6646-4156-b225-c0c7d18c39ac">

But this is definitely something we can try out :).